plugins/GSdx/GS.cpp

```diff
@@ -1086,233 +1087,123 @@ EXPORT_C GSReplay(HWND hwnd, HINSTANCE hinst, LPSTR lpszCmdLine, int nCmdShow)
 
 	::SetPriorityClass(::GetCurrentProcess(), HIGH_PRIORITY_CLASS);
 
-	if(FILE* fp = fopen(lpszCmdLine, "rb"))
-	{
-		Console console("GSdx", true);
-
-		GSinit();
+	Console console{"GSdx", true};
 
-		uint8 regs[0x2000];
-		GSsetBaseMem(regs);
+	const std::string f{lpszCmdLine};
+	const bool is_xz = f.size() >= 4 && f.compare(f.size() - 3, 3, ".xz") == 0;
 
-		s_vsync = theApp.GetConfigB("vsync");
+	auto file = is_xz
+		? std::unique_ptr<GSDumpFile>{std::make_unique<GSDumpLzma>(lpszCmdLine, nullptr)}
+		: std::unique_ptr<GSDumpFile>{std::make_unique<GSDumpRaw>(lpszCmdLine, nullptr)};
 
-		HWND hWnd = NULL;
+	GSinit();
 
-		_GSopen((void**)&hWnd, "", renderer);
+	std::array<uint8, 0x2000> regs;
+	GSsetBaseMem(regs.data());
 
-		uint32 crc;
-		fread(&crc, 4, 1, fp);
-		GSsetGameCRC(crc, 0);
+	s_vsync = theApp.GetConfigB("vsync");
 
-		GSFreezeData fd;
-		fread(&fd.size, 4, 1, fp);
-		fd.data = new uint8[fd.size];
-		fread(fd.data, fd.size, 1, fp);
-		GSfreeze(FREEZE_LOAD, &fd);
-		delete [] fd.data;
+	HWND hWnd = nullptr;
 
-		fread(regs, 0x2000, 1, fp);
+	_GSopen((void**)&hWnd, "", renderer);
 
-		long start = ftell(fp);
+	uint32 crc;
+	file->Read(&crc, 4);
+	GSsetGameCRC(crc, 0);
 
-		GSvsync(1);
+	{
+		GSFreezeData fd;
+		file->Read(&fd.size, 4);
+		std::vector<uint8> freeze_data(fd.size);
+		fd.data = freeze_data.data();
+		file->Read(fd.data, fd.size);
+		GSfreeze(FREEZE_LOAD, &fd);
+	}
 
-		struct Packet {uint8 type, param; uint32 size, addr; vector<uint8> buff;};
+	file->Read(regs.data(), 0x2000);
 
-		list<Packet*> packets;
-		vector<uint8> buff;
-		int type;
+	GSvsync(1);
 
-		while((type = fgetc(fp)) != EOF)
-		{
-			Packet* p = new Packet();
+	struct Packet {uint8 type, param; uint32 size, addr; std::vector<uint8> buff;};
 
-			p->type = (uint8)type;
+	auto read_packet = [&file](uint8 type) {
```

Comment:
Yeah, the replay stuff could probably be combined to some extent.

Personally I think it'd be nice to have a cross platform GS dump viewer at some point in the future, something that'd allow some level of automated testing.